src/lib.rs

```diff
@@ -829,11 +861,282 @@ fn assign_levels_to_removed_chars(para_level: Level, classes: &[BidiClass], leve
     }
 }
 
+/// Implementation of TextSource for UTF-8 text (a string slice).
+impl TextSource for str {
+    #[inline]
+    fn len(&self) -> usize {
+        (self as &str).len()
+    }
+    #[inline]
+    fn char_at(&self, index: usize) -> Option<(char, usize)> {
+        if index >= self.len() || !self.is_char_boundary(index) {
+            return None;
+        }
+        let ch = self[index..].chars().next().unwrap();
+        Some((ch, ch.len_utf8()))
+    }
+    #[inline]
+    fn chars(&self) -> impl DoubleEndedIterator<Item = char> {
+        (self as &str).chars()
+    }
+    #[inline]
+    fn char_indices(&self) -> impl Iterator<Item = (usize, char)> {
+        (self as &str).char_indices()
+    }
+    #[inline]
+    fn indices_lengths(&self) -> impl Iterator<Item = (usize, usize)> {
+        Utf8IndexLenIter::new(&self)
+    }
+    #[inline]
+    fn char_len(ch: char) -> usize {
+        ch.len_utf8()
+    }
+}
+
+/// Iterator over (UTF-8) string slices returning (index, char_len) tuple.
+#[derive(Debug)]
+struct Utf8IndexLenIter<'text> {
+    iter: CharIndices<'text>,
+}
+
+impl<'text> Utf8IndexLenIter<'text> {
+    #[inline]
+    pub fn new(text: &'text str) -> Self {
+        Utf8IndexLenIter {
+            iter: text.char_indices(),
+        }
+    }
+}
+
+impl Iterator for Utf8IndexLenIter<'_> {
+    type Item = (usize, usize);
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some((pos, ch)) = self.iter.next() {
+            return Some((pos, ch.len_utf8()));
+        }
+        None
+    }
+}
+
+/// Implementation of TextSource for UTF-16 text in a [u16] array.
+/// Note that there could be unpaired surrogates present!
+
+// Convenience functions to check whether a UTF16 code unit is a surrogate.
+#[inline]
+fn is_surrogate(code: u16) -> bool {
+    (code & 0xF800) == 0xD800
+}
+#[inline]
+fn is_high_surrogate(code: u16) -> bool {
+    (code & 0xFC00) == 0xD800
+}
+#[inline]
+fn is_low_surrogate(code: u16) -> bool {
+    (code & 0xFC00) == 0xDC00
+}
+
+impl TextSource for [u16] {
+    #[inline]
+    fn len(&self) -> usize {
+        (self as &[u16]).len()
+    }
+    fn char_at(&self, index: usize) -> Option<(char, usize)> {
+        if index >= self.len() {
+            return None;
+        }
+        // Get the indicated code unit; if it's not a surrogate, we're done.
+        let c = self[index];
+        if !is_surrogate(c) {
```

Comment:
nit: please just use the checked version here and don't check for surrogateyness

Reply:
Done (in following nit-fix commit)